### ISSUE TITLE
Add the ability to calculate the Jacaboian matrix to `astropy.wcs`.

### ISSUE THREAD

**Issue by odysseus9672** [astropy/astropy#10567]
### Description

A useful feature for handling coordinate transformations is the <A HREF="https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant">Jacobian matrix</A>. Primarily, it allows for a local determination of how the transformation modifies area elements (by the eigenvalues/determinant) and basis orientation (by the eigenvectors). The prototypical application is to determine in the xy coordinates which directions are North and East. at a particular point in the image. In the simple case, this would just be reporting the elements of `CDi_j` correctly packed in an array, but when distortion is relevant the local Jacobian should be modified by this. This should correctly handle all of the special cases, though, like when the header specifies two scales and an angle.

Ideally, the signatures of the functions should follow the coordinate transform functions. Something like `Jac_all_world2pix`, with the same arguments and meanings as the `all_world2pix` function, but which returns an `N` by `NAXIS` by `NAXIS` array, instead of just `N` by `NAXIS`.

### Additional context

My primary use case is overlaying text on a ds9 image and getting the text to display upright. When the fits file does unusual things with the data axis order (i.e. the Jacobian has a reflection) and/or a 90 degree rotation, it can make the text display upside-down (even when the text angle is specified relative to the fk5 coordinates as 0.0).

It would also make it more straightforward to generate compass roses to overlay on non-ds9 images.

**Comment by Maia-jp** [comment]
Hello, is this issue open?

**Comment by odysseus9672** [comment]
@Maia-jp As far as I'm concerned it is, but I have zero authority on the project.

**Comment by odysseus9672** [comment]
Most values in fits files tend to be stored in 32 bit floats. So I would recommend using `np.finfo(np.float32).eps * 8`, which is just smaller than `1e-6`, as the default delta. You can also save a couple of steps by using `np.empty((NAXIS, NAXIS))` instead of `np.eye`, since you end up overwriting it, anyway.

Honestly, this approach is probably good enough for most purposes, but I would, personally, prefer something that dug in to the wcs and calculated the Jacobian directly, without loss of significance. I don't know if that's feasible, though, considering how complicated the fits wcs standard is. The next best thing would be to bring in some package that can do automatic differentiation, like Pytorch's autograd, but that might require reworking the whole wcs package to use functions that support automatic differentiation. 

Short version: looks good enough to me.

**Comment by Maia-jp** [comment]
> I have taken the initiative to begin working on this issue. After reading the documentation for the wcs class, I have developed an implementation that can be found at the following GitHub commit: [link](https://github.com/astropy/astropy/commit/1036628ccf398a953c05155e14a3e3d5b3f36417).
> 
> I would greatly appreciate any feedback you have regarding the implementation.

@pllim I would to know your thoughts on this changes

**Comment by pllim** [comment]
I am not qualified to do WCS math, so will have to loop in @mcara . General comment: `Jac` -> `'jac` (and probably just spell it all out)

**Comment by mcara** [comment]
I do not think I understand exactly the need for this function. From the original comment:

> The prototypical application is to determine in the xy coordinates which directions are North and East. at a particular point in the image. In the simple case, this would just be reporting the elements of CDi_j correctly packed in an array, but when distortion is relevant the local Jacobian should be modified by this. This should correctly handle all of the special cases, though, like when the header specifies two scales and an angle.

 > My primary use case is overlaying text on a ds9 image and getting the text to display upright. When the fits file does unusual things with the data axis order (i.e. the Jacobian has a reflection) and/or a 90 degree rotation, it can make the text display upside-down (even when the text angle is specified relative to the fk5 coordinates as 0.0).

CD matrix is the Jacobian matrix of the transformation near CRVAL in most circumstances (I am not that familiar with all projections but I think this holds true for most if not all projections). The distortions are usually too small to cause change  in the handedness of the coordinate system. Therefore, once you have determined direction of the North, it pretty much should stay almost the same over a typical image.

If variation across the field of view is important, then please take a look at [astropy.wcs.utils.local_partial_pixel_derivatives](https://docs.astropy.org/en/stable/api/astropy.wcs.utils.local_partial_pixel_derivatives.html#astropy.wcs.utils.local_partial_pixel_derivatives). I think it might be doing what you want.

**Comment by mcara** [comment]
@Maia-jp I have tried using your function on one HST image. First, the first argument (`header`) is not used anywhere and should be removed. Then, as the first test, I tried to see if the returned Jacobian matrix would be close to the CD matrix. It was very different from the CD matrix. I think you are computing inverse of the CD matrix: `dx/d(RA)`, ...

Another aspect is that you are using `all_world2pix` which is inverse transformation of the forward transformation which is analytical. Inverse transformation is computed iteratively and, I believe, the default value for the accuracy is 1e-4. Just something to keep in mind. Probably the best value for "delta" would be something of the order of a pixel. The pixel scale  can be determined from the CD matrix (or using forward transformation `wcs_pix2world`).

However, I think [astropy.wcs.utils.local_partial_pixel_derivatives](https://docs.astropy.org/en/stable/api/astropy.wcs.utils.local_partial_pixel_derivatives.html#astropy.wcs.utils.local_partial_pixel_derivatives) does all of this and if you need the inver matrix - that is easy to invert with `numpy.linalg.inv`.
